extensions/amp-subscriptions/0.1/metering.js

```diff
@@ -0,0 +1,136 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AmpDoc as AmpDocDef} from '../../../src/service/ampdoc-impl';
+import {Services} from '../../../src/services';
+import {dev} from '../../../src/log';
+
+/** @const */
+const TAG = 'amp-subscriptions';
+
+/** @const */
+const STORAGE_KEY = 'amp-subscriptions:metering-store';
+
+/**
+ * Describes metering state for a single publication.
+ * @typedef {{
+ *   id: string,
+ *   attributes: !Array<{
+ *     name: string,
+ *     timestamp: string,
+ *   }>,
+ * }}
+ */
+let MeteringStateDef;
+
+/** Handles metering functionality. */
+export class Metering {
+  /**
+   *
+   * @param {{
+   *   ampdoc: !AmpDocDef,
+   *   platformKey: string,
+   * }} params
+   */
+  constructor({ampdoc, platformKey}) {
+    /** @private {Promise<!../../../src/service/storage-impl.Storage>} */
+    this.storagePromise_ = Services.storageForDoc(ampdoc);
+
+    /**
+     * Remembers if metering entitlements were fetched
+     * with the current metering state.
+     *
+     * This helps avoid redundant fetches.
+     * @type {boolean}
+     */
+    this.entitlementsWereFetchedWithCurrentMeteringState = false;
+
+    /**
+     * Key for the subscription platform with metering enabled.
+     * @const
+     */
+    this.platformKey = platformKey;
+  }
+
+  /**
+   * Saves metering state for a given publication ID.
+   *
+   * @param {!MeteringStateDef} meteringState
+   * @return {!Promise}
+   */
+  saveMeteringState(meteringState) {
+    const meteringStateString = JSON.stringify(meteringState);
+
+    const meteringStateChangedPromise = this.loadMeteringStateString_().then(
+      (existingMeteringStateString) =>
+        meteringStateString !== existingMeteringStateString
+    );
+
+    return meteringStateChangedPromise.then((meteringStateChanged) => {
+      if (!meteringStateChanged) {
+        // Bail. Fetching metering entitlements again is redundant,
+        // until the metering state changes.
+        return Promise.resolve();
+      }
+
+      return (
+        this.storagePromise_
```

Comment:
How will this behave if storage is not available (webview)?  presumably it should just do the double entitlement call?

Reply:
That's a good question. I just tested what happens in Chrome if all client side storage is disabled. I'm seeing exceptions when the SwG JavaScript tries to check for "sessionStorage" 
![image](https://user-images.githubusercontent.com/1216762/109170806-fa39dc00-7735-11eb-8fe6-a532f66d5352.png)

Reply:
Also the Regwall's GSI button fails to render if cookies aren't available 
![image](https://user-images.githubusercontent.com/1216762/109173352-982ea600-7738-11eb-98bf-3d5fea332ea0.png)